### PR TITLE
feat(kinesisfirehose): supports Kinesis data stream source for delivery stream

### LINKED ISSUE #10783: [aws-kinesis] Read permissions to stream doesn't include kinesis:DescribeStream
When granting an IAM role read permissions on a Kinesis stream, the resulting policy does not include the "kinesis:DescribeStream" permission, which is needed for at least Kinesis Firehose to read from it, perhaps other consumers as well.

### Reproduction Steps

```
from aws_cdk import (
    core,
    aws_kinesis as kds,
    aws_iam as iam,
)
class DFPStack(core.NestedStack):
    def __init__(
        self, scope: core.Construct, id: str, *, stream=kds.Stream, **kwargs
    ) -> None:

        self.kfh_reader = iam.Role(
            self,
            "KFHReader",
            assumed_by=iam.ServicePrincipal(service="firehose.amazonaws.com"),
        )
        self.stream = stream
        self.stream.grant_read(self.kfh_reader)
```
Results in the following output:
```
"KFHReaderDefaultPolicy98F05724": {
      "Type": "AWS::IAM::Policy",
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "kinesis:DescribeStreamSummary",
                "kinesis:GetRecords",
                "kinesis:GetShardIterator",
                "kinesis:ListShards",
                "kinesis:SubscribeToShard"
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "referencetoLongboatkinesisNestedStackkinesisNestedStackResourceFA86427BOutputsLongboatkinesisstreamDE86A4D8Arn"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "PolicyName": "KFHReaderDefaultPolicy98F05724",
        "Roles": [
          {
            "Ref": "KFHReader8A09BD1B"
          }
        ]
      },
      "Metadata": {
        "aws:cdk:path": "Longboat/DFPStack/KFHReader/DefaultPolicy/Resource"
      }
    },
```

### What did you expect to happen?

I expected the resulting IAM role and policy to actually be able to read from the supplied Kinesis stream when used by Kinesis Firehose. This would require the "kinesis:DescribeStream" permission to be given.

### What actually happened?

Stack creation fails, because the "kinesis:DescribeStream" permission is lacking.

Cloudformation gives the following error:
`Role arn:aws:iam::<account>:role/<role_name> is not authorized to perform: kinesis:DescribeStream on resource arn:aws:kinesis:eu-west-1:<account>:stream/<stream_name>. (Service: Firehose, Status Code: 400, Request ID: <request_id>, Extended Request ID: <extended_request_id>`

### Environment

  - **CLI Version:** 1.66.0 (build 459488d)
  - **Framework Version:**  1.66.0
  - **Node.js Version:**  v12.19.0
  - **OS:** KDE Neon 5.19 (Ubuntu 20.04 focal)
  - **Language (Version):**  Python 3.8.5

### Other

--- 

This is :bug: Bug Report

### LINKED ISSUE #15500: (aws-kinesisfirehose): Kinesis source stream for DeliveryStream
Allow users to specify a Kinesis data stream as the source for their `DeliveryStream`.

Part of https://github.com/aws/aws-cdk/issues/7536

### Use Case

Linking Kinesis data streams as the source for a Kinesis Data Firehose delivery stream. 

- [x] ensure we have handled the race condition in https://github.com/aws/aws-cdk/issues/10783#issuecomment-706589031

### Proposed Solution

See [RFC 340: Kinesis Firehose Delivery Stream](https://github.com/aws/aws-cdk-rfcs/pull/342)

And [this branch with the prototype](https://github.com/madeline-k/aws-cdk/tree/madeline-k/feat/kinesisfirehose-l2)

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by BenChaimberg**
closes #15500 
closes #10783 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-kinesisfirehose/lib/delivery-stream.ts [hunk 1]
```diff
@@ -196,6 +197,13 @@ export interface DeliveryStreamProps {
    */
   readonly deliveryStreamName?: string;
 
+  /**
+   * The Kinesis data stream to use as a source for this delivery stream.
+   *
+   * @default - data is written to the delivery stream via a direct put.
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose/lib/delivery-stream.ts:203]:
```suggestion
   * @default - data must be written to the delivery stream via a direct put.
```

File: packages/@aws-cdk/aws-kinesisfirehose/lib/delivery-stream.ts [hunk 2]
```diff
@@ -319,6 +327,12 @@ export class DeliveryStream extends DeliveryStreamBase {
     });
     this.grantPrincipal = role;
 
+    if (
+      props.sourceStream &&
+        (props.encryption === StreamEncryption.AWS_OWNED || props.encryption === StreamEncryption.CUSTOMER_MANAGED || props.encryptionKey)
+    ) {
+      throw new Error('Requested server-side encryption but delivery stream source is a Kinesis Data Stream. Specify server-side encryption on the Data Stream instead.');
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose/lib/delivery-stream.ts:334]: For consistency:
```suggestion
      throw new Error('Requested server-side encryption but delivery stream source is a Kinesis data stream. Specify server-side encryption on the data stream instead.');
```

The alternative would be keeping the capitalization, and adding it elsewhere.

File: packages/@aws-cdk/aws-kinesisfirehose/lib/delivery-stream.ts [hunk 3]
```diff
@@ -339,15 +353,29 @@ export class DeliveryStream extends DeliveryStreamBase {
      */
     encryptionKey?.grantEncryptDecrypt(role);
 
+    const sourceStreamConfig = props.sourceStream ? {
+      kinesisStreamArn: props.sourceStream.streamArn,
+      roleArn: role.roleArn,
+    } : undefined;
+    const readStreamGrant = props.sourceStream?.grantRead(role);
+    if (readStreamGrant && readStreamGrant.principalStatement) {
+      readStreamGrant.principalStatement.addActions('kinesis:DescribeStream');
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose/lib/delivery-stream.ts:362]: Can you add a comment about why this is necessary? Looks like this fixes the issue described in  https://github.com/aws/aws-cdk/issues/10783?

File: packages/@aws-cdk/aws-kinesisfirehose/test/integ.delivery-stream.source-stream.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    const bucketGrant = bucket.grantReadWrite(role);
+    return {
+      extendedS3DestinationConfiguration: {
+        bucketArn: bucket.bucketArn,
+        roleArn: role.roleArn,
+      },
+      dependables: [bucketGrant],
+    };
+  },
+};
```
Comment (madeline-k) [inline packages/@aws-cdk/aws-kinesisfirehose/test/integ.delivery-stream.source-stream.ts:32]: Is this duplicated with the other integration test? Can you make one shared construct for this for the unit and integ tests?

**Review by madeline-k — COMMENTED** [review]
This looks great! A couple minor comments.

File: packages/@aws-cdk/aws-kinesisfirehose/test/integ.delivery-stream.source-stream.ts (hunk 4 above)
Comment (BenChaimberg) [reply packages/@aws-cdk/aws-kinesisfirehose/test/integ.delivery-stream.source-stream.ts:32]: I think for literacy purposes, I'd rather keep this duplicated

**Review by madeline-k — APPROVED** [review]
Looks great 👍

**Pull request merged**